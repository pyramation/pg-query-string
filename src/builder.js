import squel from './squel';
import escape from 'pg-escape';

// 0 = pg
// 1 = pg-promise

export class Query {
  constructor({ client, type = 0 }) {
    this.client = client;
    this.type = type;
  }
  async query(asParamsObj) {
    if (!this.type) {
      return await this.client.query(asParamsObj);
    } else {
      return await this.client.any(asParamsObj.type, asParamsObj.values);
    }
  }
}

export class Schema {
  constructor(name) {
    this.name = name;
  }
  table(name, structure) {
    return new Table(name, structure, this.name);
  }
  function(name, structure) {
    return new Func(name, structure, this.name);
  }
}

export class Table {
  constructor(name, structure = {}, schema) {
    if (name.includes('.')) {
      const parts = name.split('.');
      name = parts[1];
      schema = parts[0];
    }
    this.name = name;
    this.qualified = schema ? escape('%I.%I', schema, name) : name;

    this.structure = Object.entries(structure).reduce((m, [k, v]) => {
      m[k] = {
        type: v,
        isArray: v.replace(/[\s]/g, '').endsWith('[]')
      };
      return m;
    }, {});
  }
  insert(opts, returning = ['*']) {
    const builder = Object.entries(opts).reduce((m, v) => {
      const [field, value] = v;
      if (this.structure[field]) {
        if (this.structure[field].isArray) {
          return m.set(field, [value], this.structure[field].type);
        } else {
          return m.set(field, value, this.structure[field].type);
        }
      } else {
        return m.set(field, value);
      }
    }, squel.insert().into(this.qualified));
    if (returning && returning.length) {
      return builder.returning(returning.join(',')).toParam();
    }
    return builder.toParam();
  }
  update(fields, condition = {}) {
    const casts = [];
    const values = [];

    const builder = Object.entries(fields).reduce((m, v) => {
      const [field, value] = v;
      casts.push(this.structure[field]?.type);
      values.push(value);

      if (this.structure[field]) {
        return m.set(field, value, this.structure[field].type);
      } else {
        return m.set(field, value);
      }
    }, squel.update().table(this.qualified));

    const entries = Object.entries(condition);
    if (!entries.length) return builder.toParam();

    const where = entries.reduce((m, v) => {
      const [field, value] = v;
      casts.push(this.structure[field]?.type);
      values.push(value);
      return m.and(`${field} = ?`);
    }, squel.expr());

    const params = builder.where(where).toParam();

    const str = values.reduce((m, v, i) => {
      const n = i + 1;
      if (typeof casts[i] === 'undefined') return m;
      // if it's already cast, leave it alone...
      const reg = new RegExp(`\\$${n}::`);
      if (reg.test(m)) return m;
      // cast it
      return m.replace('$' + n, '$' + n + '::' + casts[i]);
    }, params.text);

    return {
      text: str,
      values
    };
  }
  selectOne(fields = [], opts = {}) {
    return this.select(fields, opts, 1);
  }
  select(fields = [], opts = {}, limit = 0) {
    const selection = fields.reduce((m, v) => {
      if (v === '*') return m;
      return m.field(v);
    }, squel.select().from(this.qualified));

    const entries = Object.entries(opts);

    const builder = limit > 0 ? selection.limit(limit) : selection;

    if (!entries.length) return builder.toParam();

    const casts = [];
    const values = [];

    const where = entries.reduce((m, v) => {
      const [field, value] = v;
      casts.push(this.structure[field]?.type);
      values.push(value);
      return m.and(`${field} = ?`);
    }, squel.expr());

    if (limit > 0 && builder.toParam().values.length) {
      casts.push();
      values.push(builder.toParam().values[0]);
    }

    const a = builder.where(where).toParam();

    const str = values.reduce((m, v, i) => {
      const n = i + 1;
      if (typeof casts[i] === 'undefined') return m;
      return m.replace('$' + n, '$' + n + '::' + casts[i]);
    }, a.text);

    return {
      text: str,
      values
    };
  }
  delete(opts = {}, limit = 0) {
    const deletion = squel.delete().from(this.qualified);

    const entries = Object.entries(opts);

    const builder = limit > 0 ? deletion.limit(limit) : deletion;

    if (!entries.length) return builder.toParam();

    const casts = [];
    const values = [];

    const where = entries.reduce((m, v) => {
      const [field, value] = v;
      casts.push(this.structure[field]?.type);
      values.push(value);
      return m.and(`${field} = ?`);
    }, squel.expr());

    if (limit > 0 && builder.toParam().values.length) {
      casts.push();
      values.push(builder.toParam().values[0]);
    }

    const a = builder.where(where).toParam();

    const str = values.reduce((m, v, i) => {
      const n = i + 1;
      if (typeof casts[i] === 'undefined') return m;
      return m.replace('$' + n, '$' + n + '::' + casts[i]);
    }, a.text);

    return {
      text: str,
      values
    };
  }
}

export class Func {
  constructor(name, structure = {}, schema) {
    if (name.includes('.')) {
      const parts = name.split('.');
      name = parts[1];
      schema = parts[0];
    }
    this.name = name;
    this.qualified = schema ? escape('%I.%I', schema, name) : name;

    this.structure = Object.entries(structure).reduce((m, [k, v]) => {
      m[k] = {
        type: v,
        isArray: v.replace(/[\s]/g, '').endsWith('[]')
      };
      return m;
    }, {});
  }
  call(fields = [], params) {
    if (!Array.isArray(fields)) {
      params = fields;
      fields = [];
    }

    const entries = Object.entries(params);
    const casts = [];
    const values = [];
    entries.forEach(v => {
      const [field, value] = v;
      casts.push(this.structure[field]?.type);
      values.push(value);
    });

    const builder = fields.reduce((m, v) => {
      if (v === '*') return m;
      return m.field(v);
    }, squel.select().from(`${this.qualified}(${values.map(v => '?').join(',')})`));

    const a = builder.toParam();

    const str = values.reduce((m, v, i) => {
      const n = i + 1;
      if (typeof casts[i] === 'undefined') return m;
      return m.replace('$' + n, '$' + n + '::' + casts[i]);
    }, a.text);

    return {
      text: str,
      values
    };
  }
}
