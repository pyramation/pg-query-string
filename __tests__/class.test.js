import { Table, Schema } from '../src';
import cases from 'jest-in-case';

const table = new Table('users', {
  id: 'uuid',
  username: 'text',
  email: 'email',
  password: 'password',
  active: 'boolean',
  github: 'githublink',
  facebook: 'fblink',
  myspace: 'myslink'
});

const schema1 = new Schema('myschema');
const schema2 = new Schema('my-schema');
const table2 = schema1.table('users');
const table3 = schema2.table('users');

it('select', () => {
  expect(
    table.select(['id', 'email'], {
      username: 'pyramation',
      email: 'pyramation@gmail.com'
    })
  ).toMatchSnapshot();
  expect(
    table2.select(['id', 'email'], {
      username: 'pyramation',
      email: 'pyramation@gmail.com'
    })
  ).toMatchSnapshot();
  expect(
    table3.select(['id', 'email'], {
      username: 'pyramation',
      email: 'pyramation@gmail.com'
    })
  ).toMatchSnapshot();
});

it('selectOne', () => {
  expect(
    table.selectOne(['id', 'email'], {
      username: 'pyramation',
      email: 'pyramation@gmail.com'
    })
  ).toMatchSnapshot();
});

it('selectOne by id', () => {
  expect(
    table.selectOne([], {
      id: 'boom'
    })
  ).toMatchSnapshot();
});

it('insert', () => {
  expect(
    table.insert({ username: 'pyramation', email: 'pyramation@gmail.com' })
  ).toMatchSnapshot();
  expect(
    table.insert(
      { username: 'pyramation', email: 'pyramation@gmail.com' },
      false
    )
  ).toMatchSnapshot();
  expect(
    table.insert({ username: 'pyramation', email: 'pyramation@gmail.com' }, [
      'id',
      'username'
    ])
  ).toMatchSnapshot();
});
it('update', () => {
  expect(
    table.update(
      {
        username: 'pyramation',
        email: 'pyramation@gmail.com'
      },
      {
        email: 'pyramation@gmail.com'
      }
    )
  ).toMatchSnapshot();
});

it('update many fields', () => {
  expect(
    table.update(
      {
        email: 'cheerios@gmail.com',
        username: 'pyramation',
        active: true,
        password: 'secure',
        avatar: 'http://some.link',
        github: 'pyramation',
        launchql: 'yes',
        facebook: 'danlynch',
        myspace: 'forgot',
        tags: 'pyramation'
      },
      {
        email: 'pyramation@gmail.com'
      }
    )
  ).toMatchSnapshot();
});

it('delete via a class', () => {
  expect(
    table.delete({ username: 'pyramation', email: 'pyramation@gmail.com' })
  ).toMatchSnapshot();
  expect(
    table.delete({ username: 'pyramation', email: 'pyramation@gmail.com' }, 2)
  ).toMatchSnapshot();
});

it('test', () => {
  const schema = new Schema('public');
  const table = schema.table('users', {
    id: 'uuid',
    username: 'text',
    email: 'email',
    password: 'password',
    active: 'boolean',
    tags: 'text[]'
  });

  expect(
    table.update(
      {
        active: true,
        tags: ['pyramation']
      },
      {
        id: 1
      }
    )
  ).toMatchSnapshot();
});
