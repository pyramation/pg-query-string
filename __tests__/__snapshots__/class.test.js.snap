// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`delete via a class 1`] = `
Object {
  "text": "DELETE FROM users WHERE (username = $1::text AND email = $2::email)",
  "values": Array [
    "pyramation",
    "pyramation@gmail.com",
  ],
}
`;

exports[`delete via a class 2`] = `
Object {
  "text": "DELETE FROM users WHERE (username = $1::text AND email = $2::email) LIMIT $3",
  "values": Array [
    "pyramation",
    "pyramation@gmail.com",
    2,
  ],
}
`;

exports[`insert 1`] = `
Object {
  "text": "INSERT INTO users (username, email) VALUES ($1::text, $2::email) RETURNING *",
  "values": Array [
    "pyramation",
    "pyramation@gmail.com",
  ],
}
`;

exports[`insert 2`] = `
Object {
  "text": "INSERT INTO users (username, email) VALUES ($1::text, $2::email)",
  "values": Array [
    "pyramation",
    "pyramation@gmail.com",
  ],
}
`;

exports[`insert 3`] = `
Object {
  "text": "INSERT INTO users (username, email) VALUES ($1::text, $2::email) RETURNING id,username",
  "values": Array [
    "pyramation",
    "pyramation@gmail.com",
  ],
}
`;

exports[`insert 4`] = `
Object {
  "text": "INSERT INTO users (username, email, friend_ids) VALUES ($1::text, $2::email, ($3::uuid[])) RETURNING id,username",
  "values": Array [
    "pyramation",
    "pyramation@gmail.com",
    Array [
      "a",
      "b",
    ],
  ],
}
`;

exports[`select 1`] = `
Object {
  "text": "SELECT id, email FROM users WHERE (username = $1::text AND email = $2::email)",
  "values": Array [
    "pyramation",
    "pyramation@gmail.com",
  ],
}
`;

exports[`select 2`] = `
Object {
  "text": "SELECT id, email FROM myschema.users WHERE (username = $1 AND email = $2)",
  "values": Array [
    "pyramation",
    "pyramation@gmail.com",
  ],
}
`;

exports[`select 3`] = `
Object {
  "text": "SELECT id, email FROM \\"my-schema\\".users WHERE (username = $1 AND email = $2)",
  "values": Array [
    "pyramation",
    "pyramation@gmail.com",
  ],
}
`;

exports[`selectOne 1`] = `
Object {
  "text": "SELECT id, email FROM users WHERE (username = $1::text AND email = $2::email) LIMIT $3",
  "values": Array [
    "pyramation",
    "pyramation@gmail.com",
    1,
  ],
}
`;

exports[`selectOne by id 1`] = `
Object {
  "text": "SELECT * FROM users WHERE (id = $1::uuid) LIMIT $2",
  "values": Array [
    "boom",
    1,
  ],
}
`;

exports[`test 1`] = `
Object {
  "text": "UPDATE public.users SET active = $1::boolean, tags = ($2::text[]) WHERE (id = $3::uuid)",
  "values": Array [
    true,
    Array [
      "pyramation",
    ],
    1,
  ],
}
`;

exports[`update 1`] = `
Object {
  "text": "UPDATE users SET username = $1::text, email = $2::email WHERE (email = $3::email)",
  "values": Array [
    "pyramation",
    "pyramation@gmail.com",
    "pyramation@gmail.com",
  ],
}
`;

exports[`update many fields 1`] = `
Object {
  "text": "UPDATE users SET email = $1::email, username = $2::text, active = $3::boolean, password = $4::password, avatar = $5, github = $6::githublink, launchql = $7, facebook = $8::fblink, myspace = $9::myslink, tags = $10 WHERE (email = $11::email)",
  "values": Array [
    "cheerios@gmail.com",
    "pyramation",
    true,
    "secure",
    "http://some.link",
    "pyramation",
    "yes",
    "danlynch",
    "forgot",
    "pyramation",
    "pyramation@gmail.com",
  ],
}
`;
